src/utils/staking/index.ts

```diff
@@ -43,3 +63,133 @@ export const buildStakingOutput = (
   }
   return psbtOutputs;
 };
+
+/**
+ * Get the staking term, it will ignore the `stakingTimeBlocks` and
+ * use the value from params if the min and max staking time blocks are the same
+ *
+ * @param {Params} params - The staking parameters.
+ * @param {number} term - The staking term.
+ * @returns {number} - The staking term.
+ */
+export const getStakingTerm = (params: Params, term: number) => {
+  // check if term is fixed
+  let termWithFixed;
+  if (params.minStakingTimeBlocks === params.maxStakingTimeBlocks) {
+    // if term is fixed, use the API value
+    termWithFixed = params.minStakingTimeBlocks;
+  } else {
+    // if term is not fixed, use the term from the input
+    termWithFixed = term;
+  }
+  return termWithFixed;
+};
+
+/**
+ * Validate the staking transaction input data.
+ *
+ * @param {number} stakingAmountSat - The staking amount in satoshis.
+ * @param {number} stakingTerm - The staking term in blocks.
+ * @param {Params} params - The staking parameters.
+ * @param {UTXO[]} inputUTXOs - The input UTXOs.
+ * @param {number} feeRate - The Bitcoin fee rate in sat/vbyte
+ * @throws {StakingError} - If the input data is invalid.
+ */
+export const validateStakingTxInputData = (
+  stakingAmountSat: number,
+  stakingTerm: number,
+  params: Params,
+  inputUTXOs: UTXO[],
+  feeRate: number,
+) => {
+  if (
+    stakingAmountSat < params.minStakingAmountSat ||
+    stakingAmountSat > params.maxStakingAmountSat
+  ) {
+    throw new StakingError(
+      StakingErrorCode.INVALID_INPUT, "Invalid staking amount",
+    );
+  }
+
+  if (
+    stakingTerm < params.minStakingTimeBlocks ||
+    stakingTerm > params.maxStakingTimeBlocks
+  ) {
+    throw new StakingError(
+      StakingErrorCode.INVALID_INPUT, "Invalid staking term",
+    );
+  }
+
+  if (inputUTXOs.length == 0) {
+    throw new StakingError(
+      StakingErrorCode.INVALID_INPUT, "No input UTXOs provided",
+    );
+  } else if (feeRate <= 0) {
+    throw new StakingError(
+      StakingErrorCode.INVALID_INPUT, "Invalid fee rate",
+    );
+  }
+}
+
+/**
+ * Validate the parameters for staking.
+ * 
+ * @param {Params} params - The staking parameters.
+ * @throws {StakingError} - If the parameters are invalid.
+ */
+export const validateParams = (params: Params) => {
+  if (params.covenantPks.length == 0) {
```

Comment:
We should validate that the covenant keys are valid keys without a coordinate

Reply:
> The application layer should have as little knowledge as possible about blockchain-specific details.

The application layer is responsible for parsing the global parameters and passing a unified format to the library. Why should the library be responsible for parsing or knowing about the idiosyncrasies of the global parameters. This is a low level library concerned with staking.

> Why should React applications need to know the intricacies of how BTC scripting works, particularly regarding coordinates?

These are not intricacies of how BTC scripting works, they refer to the Bitcoin public keys which are tightly related to the wallet. The application should know about what data it should get from the wallet and what data it should pass to the low level library which does not care where the data is coming from.

Reply:
I think the right thing to do is work with only no coordinate public keys. I don't see much of a reason for supporting multiple types of keys as the library only seems to need keys without coordinates. I don't expect we'll need such support in the future, and if we do, we should do it in a more controlled way with wrapper methods that invoke the internal ones with strict requirements -- current approach of having our methods get two types of keys (even in the same function call) is confusing and prone to errors.

@KonradStaniec what's your opinion?

Reply:
So in my perfect work would not accept `strings` at library level, but some well defined domain type representing public key. So `  covenantPks: string[];` should be rather `  covenantPks: PublicKey[];` as not only length determine whether public key is valid. (for example it must also be point on secp256k1 curve)

If having separate domain types is not an option, I would probably restrict keys to be valid at the boundaries of the library i.e we are building library which builds tapscripts and valid keys in tapscript must have 32bytes.

So if someone passes 33byte key, one if the possiblites is that he thought that we are building standard script in which while key is used.

Reply:
>@KonradStaniec It’s definitely an option, and we actually created a ticket for this a few weeks ago: https://github.com/babylonlabs-io/btc-staking-ts/issues/26. However, I’m not keen on implementing everything in a single PR

Agreed on not having everything in single pr 👍 

>In general, I like the approach you mentioned of using a custom type. It’s essentially a wrapper type around the string that includes validation and utility methods for constructing the key from a string. This is similar to the proposed string solution, it allows the application layer to avoid dealing directly with BTC-specific key operations— but in a much cleaner and nicer way.

yup that is the case. Ideally such custom type would have:
- some smart constructors which makes it impossible or at least hard to construct such type from incorrect data
- helper methods for whatever functionalities we need
- in this way library do not need to use simple types like `string` and `[]byte` and treat every `PublicKey` as valid public key